# Tutorial 1B Hands on data - wrangling data with Python & pandas



In two parts:
* reproducing the DataWrangler process (using the same 'Air crashes' data) and 
* bad, bad data investigations


** Before you start, a most important thing to do, check your python version**

In [1]:
import sys
print (sys.version_info)

sys.version_info(major=3, minor=6, micro=4, releaselevel='final', serial=0)


In [2]:
import pandas as pd # not the bamboo eating bear... 'Panel Data' 

# Part 1 
## Step 1 Read data 

In [3]:
# load data
df = pd.read_csv('AirCrashes.csv') # df is a dataframe, confirm with: type(df)
df.shape

(927, 1)

#### How many lines of data?
#### How many did you get with DataWrangler?

Have a look a the first few rows:

In [4]:
df.head()

Incident American Airlines Flight 11 involving a Boeing 767-223ER in 2001
Casualties                                     Extremely High                       
Total Dead                                               1692                       
Crew                                                       11                       
Passengers                                                 81                       
Ground                                                   1600

That's not quite right.. the first line has been stolen for the title/header
#### Does read_csv ignore empty lines? 
e.g. line 18 should be blank

skip_blank_lines = True (the default) see:

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html

Try again, force our own headers upon the data using default column names from DataWrangler (split/extract etc.) 

This pushes the first incident down into the data where it belongs (but adds a NaN, below)

In [5]:
df = pd.read_csv('AirCrashes.csv', names = ['split', 'split1']) 
# column names (split, split1) replicate DataWrangler column names

df.head(10)

split                    split1
0  Incident American Airlines Flight 11 involving...                       NaN
1                                         Casualties            Extremely High
2                                         Total Dead                      1692
3                                               Crew                        11
4                                         Passengers                        81
5                                             Ground                      1600
6                                              Notes              No survivors
7                                               Type                       INH
8                                             Reason                    Attack
9                                           Location  New York - New York - US

## Step 2 Extract index(s)

### Extract flight information
The flight information is between "Incident" and "involving" in the "split" cell.
Now, extract flights based on one of the suggestions from DataWrangler, i.e., "(Incident (.*) involving)".

In [6]:
# treat the 'split' column as a str, then 
# use the extract method on the str
flight_no = df['split'].str.extract("Incident (.*) involving")

/home/zby0902/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:3: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  app.launch_new_instance()


### That seems to have worked...

We got 'American Airlines Flight 11' and 'United Airlines Flight 175' etc but lost all the other data and gained a bunch of NaNs

#### What is this 'str.extract' code anyway? (add a comment to decode or explain it)
>**Answer** To extract substring from each cell in this roll that mathches the pattern of the regular expression
extract("Incident (.*) involving")


### But we want planes in a new column, we can create one called 'extract' like this:

df['extract'] = df['split'].str.extract("Incident (.*) involving")

But this would be the last column, we want it in the second (location is not critical but it can be done so why not). 

Now, use the DataFrame's insert function.

In [7]:
df.insert(loc=1,column='extract',value=flight_no)

In [8]:
df.head(20)

split  \
0   Incident American Airlines Flight 11 involving...   
1                                          Casualties   
2                                          Total Dead   
3                                                Crew   
4                                          Passengers   
5                                              Ground   
6                                               Notes   
7                                                Type   
8                                              Reason   
9                                            Location   
10                                            Country   
11                                              Phase   
12                                               Date   
13                                           Latitude   
14                                          Longitude   
15                                      Circumstances   
16  Incident United Airlines Flight 175 involving ...   
17                                         Casualties   
18                                         Total Dead   
19                                               Crew   

                        extract                    split1  
0   American Airlines Flight 11                       NaN  
1                           NaN            Extremely High  
2                           NaN                      1692  
3                           NaN                        11  
4                           NaN                        81  
5                           NaN                      1600  
6                           NaN              No survivors  
7                           NaN                       INH  
8                           NaN                    Attack  
9                           NaN  New York - New York - US  
10                          NaN                        US  
11                          NaN                       ENR  
12                          NaN                2001-09-11  
13                          NaN                40.7143528  
14                          NaN               -74.0059731  
15                          NaN    Good Visibility by Day  
16   United Airlines Flight 175                       NaN  
17                          NaN            Extremely High  
18                          NaN                       965  
19                          NaN                         9

So there's the flight information in it's own column, plus a whole lot of NaNs

We could replace all the NaN with spaces or similar but they can wait

Now we want the aircraft in it's own column, similar to above, 
based on the suggestion from DataWrangler, note spaces in "\ a (.*)\ in "

#### but is this optimal?  

In [9]:
# and repeat to get the aircraft type that appears between "a" and "in"
aircraft_type = df['split'].str.extract("\ a (.*)\ in ")
df.insert(loc=2,column='extract1',value=aircraft_type)
# and df.head(20) to confirm
df.head(20)

/home/zby0902/anaconda3/lib/python3.6/site-packages/ipykernel/__main__.py:2: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  from ipykernel import kernelapp as app


split  \
0   Incident American Airlines Flight 11 involving...   
1                                          Casualties   
2                                          Total Dead   
3                                                Crew   
4                                          Passengers   
5                                              Ground   
6                                               Notes   
7                                                Type   
8                                              Reason   
9                                            Location   
10                                            Country   
11                                              Phase   
12                                               Date   
13                                           Latitude   
14                                          Longitude   
15                                      Circumstances   
16  Incident United Airlines Flight 175 involving ...   
17                                         Casualties   
18                                         Total Dead   
19                                               Crew   

                        extract          extract1                    split1  
0   American Airlines Flight 11  Boeing 767-223ER                       NaN  
1                           NaN               NaN            Extremely High  
2                           NaN               NaN                      1692  
3                           NaN               NaN                        11  
4                           NaN               NaN                        81  
5                           NaN               NaN                      1600  
6                           NaN               NaN              No survivors  
7                           NaN               NaN                       INH  
8                           NaN               NaN                    Attack  
9                           NaN               NaN  New York - New York - US  
10                          NaN               NaN                        US  
11                          NaN               NaN                       ENR  
12                          NaN               NaN                2001-09-11  
13                          NaN               NaN                40.7143528  
14                          NaN               NaN               -74.0059731  
15                          NaN               NaN    Good Visibility by Day  
16   United Airlines Flight 175    Boeing 767-222                       NaN  
17                          NaN               NaN            Extremely High  
18                          NaN               NaN                       965  
19                          NaN               NaN                         9

Drop the empty lines 

In [10]:
df['split'].duplicated();

In [11]:
df = df.dropna(how='all') 
df

split  \
0    Incident American Airlines Flight 11 involving...   
1                                           Casualties   
2                                           Total Dead   
3                                                 Crew   
4                                           Passengers   
5                                               Ground   
6                                                Notes   
7                                                 Type   
8                                               Reason   
9                                             Location   
10                                             Country   
11                                               Phase   
12                                                Date   
13                                            Latitude   
14                                           Longitude   
15                                       Circumstances   
16   Incident United Airlines Flight 175 involving ...   
17                                          Casualties   
18                                          Total Dead   
19                                                Crew   
20                                          Passengers   
21                                              Ground   
22                                               Notes   
23                                                Type   
24                                              Reason   
25                                            Location   
26                                             Country   
27                                               Phase   
28                                                Date   
29                                            Latitude   
..                                                 ...   
898                                         Total Dead   
899                                               Crew   
900                                         Passengers   
901                                             Ground   
902                                              Notes   
903                                               Type   
904                                             Reason   
905                                           Location   
906                                            Country   
907                                              Phase   
908                                               Date   
909                                           Latitude   
910                                          Longitude   
911                                      Circumstances   
912  Incident American Airlines Flight 965 involvin...   
913                                         Casualties   
914                                         Total Dead   
915                                               Crew   
916                                         Passengers   
917                                             Ground   
918                                              Notes   
919                                               Type   
920                                             Reason   
921                                           Location   
922                                            Country   
923                                              Phase   
924                                               Date   
925                                           Latitude   
926                                          Longitude   
927                                      Circumstances   

                          extract          extract1                    split1  
0     American Airlines Flight 11  Boeing 767-223ER                       NaN  
1                             NaN               NaN            Extremely High  
2                             NaN               NaN                      1692  
3                             NaN               NaN                        11  
4                             NaN               NaN 

## Step 3  'Fill down'

We want to 'fill down' the indexes  (e.g. lines 1 to 15 should be associated with line 0)

There are several options
* na.locf() method from zoo package. 
* ddply() from plyr
* bfill()
* fillna()


In [12]:
# magic, take the previous value (not NaN) and fill down
df = df.fillna(method='pad')
# http://pandas.pydata.org/pandas-docs/stable/missing_data.html#filling-missing-values-fillna

In [13]:
df.head(200)

split  \
0    Incident American Airlines Flight 11 involving...   
1                                           Casualties   
2                                           Total Dead   
3                                                 Crew   
4                                           Passengers   
5                                               Ground   
6                                                Notes   
7                                                 Type   
8                                               Reason   
9                                             Location   
10                                             Country   
11                                               Phase   
12                                                Date   
13                                            Latitude   
14                                           Longitude   
15                                       Circumstances   
16   Incident United Airlines Flight 175 involving ...   
17                                          Casualties   
18                                          Total Dead   
19                                                Crew   
20                                          Passengers   
21                                              Ground   
22                                               Notes   
23                                                Type   
24                                              Reason   
25                                            Location   
26                                             Country   
27                                               Phase   
28                                                Date   
29                                            Latitude   
..                                                 ...   
170                                            Country   
171                                              Phase   
172                                               Date   
173                                           Latitude   
174                                          Longitude   
175                                      Circumstances   
176  Incident Pan Am Flight 103 involving a Boeing ...   
177                                         Casualties   
178                                         Total Dead   
179                                               Crew   
180                                         Passengers   
181                                             Ground   
182                                              Notes   
183                                               Type   
184                                             Reason   
185                                           Location   
186                                            Country   
187                                              Phase   
188                                               Date   
189                                           Latitude   
190                                          Longitude   
191                                      Circumstances   
192  Incident Korean Air Lines Flight 007 involving...   
193                                         Casualties   
194                                         Total Dead   
195                                               Crew   
196                                         Passengers   
197                                             Ground   
198                                              Notes   
199                                               Type   

                          extract                    extract1  \
0     American Airlines Flight 11            Boeing 767-223ER   
1     American Airlines Flight 11            Boeing 767-223ER   
2     American Airlines Flight 11            Boeing 767-223ER   
3     American Airlines Flight 11            Boeing 767-223ER   
4     American Airlines Flight 11            Boeing 767-223ER   
5     American Airlines Flight 11            Boeing 767-223ER 

## Step 4 Remove the index row

We need to delete all the 'incident' rows, they have served their purpose and are now redundant. 

In [14]:
df[df['split'].str.contains('Incident')].index

Int64Index([  0,  16,  32,  48,  64,  80,  96, 112, 128, 144, 160, 176, 192,
            208, 224, 240, 256, 272, 288, 304, 320, 336, 352, 368, 384, 400,
            416, 432, 448, 464, 480, 496, 512, 528, 544, 560, 576, 592, 608,
            624, 640, 656, 672, 688, 704, 720, 736, 752, 768, 784, 800, 816,
            832, 848, 864, 880, 896, 912],
           dtype='int64')

In [15]:
# use str.contain function to get the row index.
# keep everything that doesn't have "Incident" in it 
df = df.drop(df[df['split'].str.contains('Incident')].index)
# do we need to worry about a plane called "Incident" or "Incident weather" etc???

In [16]:
# and check with e.g. df.shape 
print(df.shape)
df

(870, 4)


split                               extract          extract1  \
1       Casualties           American Airlines Flight 11  Boeing 767-223ER   
2       Total Dead           American Airlines Flight 11  Boeing 767-223ER   
3             Crew           American Airlines Flight 11  Boeing 767-223ER   
4       Passengers           American Airlines Flight 11  Boeing 767-223ER   
5           Ground           American Airlines Flight 11  Boeing 767-223ER   
6            Notes           American Airlines Flight 11  Boeing 767-223ER   
7             Type           American Airlines Flight 11  Boeing 767-223ER   
8           Reason           American Airlines Flight 11  Boeing 767-223ER   
9         Location           American Airlines Flight 11  Boeing 767-223ER   
10         Country           American Airlines Flight 11  Boeing 767-223ER   
11           Phase           American Airlines Flight 11  Boeing 767-223ER   
12            Date           American Airlines Flight 11  Boeing 767-223ER   
13        Latitude           American Airlines Flight 11  Boeing 767-223ER   
14       Longitude           American Airlines Flight 11  Boeing 767-223ER   
15   Circumstances           American Airlines Flight 11  Boeing 767-223ER   
17      Casualties            United Airlines Flight 175    Boeing 767-222   
18      Total Dead            United Airlines Flight 175    Boeing 767-222   
19            Crew            United Airlines Flight 175    Boeing 767-222   
20      Passengers            United Airlines Flight 175    Boeing 767-222   
21          Ground            United Airlines Flight 175    Boeing 767-222   
22           Notes            United Airlines Flight 175    Boeing 767-222   
23            Type            United Airlines Flight 175    Boeing 767-222   
24          Reason            United Airlines Flight 175    Boeing 767-222   
25        Location            United Airlines Flight 175    Boeing 767-222   
26         Country            United Airlines Flight 175    Boeing 767-222   
27           Phase            United Airlines Flight 175    Boeing 767-222   
28            Date            United Airlines Flight 175    Boeing 767-222   
29        Latitude            United Airlines Flight 175    Boeing 767-222   
30       Longitude            United Airlines Flight 175    Boeing 767-222   
31   Circumstances            United Airlines Flight 175    Boeing 767-222   
..             ...                                   ...               ...   
897     Casualties  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
898     Total Dead  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
899           Crew  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
900     Passengers  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
901         Ground  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
902          Notes  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
903           Type  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
904         Reason  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
905       Location  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
906        Country  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
907          Phase  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
908           Date  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
909       Latitude  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
910      Longitude  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
911  Circumstances  China Northwest Airlines Flight 2303   Tupolev Tu-154M   
913     Casualties          American Airlines Flight 965    Boeing Tu-154M   
914     Total Dead          American Airlines Flight 965    Boeing Tu-154M   
915           Crew          American Airlines Flight 965    Boeing Tu-154M   
916     Passengers          American Airlines Flight 965    Boeing Tu-154M   
917         Ground          America

## Now to 'unfold', there are several options

* melt()
* stack, unstack?
* pivot_table()
* pivot()

http://pandas.pydata.org/pandas-docs/stable/reshaping.html

In [17]:
data = df.pivot('extract', 'split', 'split1') 
# the parameters above are (left to right) index, columns, values 
# You can also write it like this:
#data = df.pivot(index = 'extract', columns = 'split', values = 'split1') 

In [18]:
data.shape 

(58, 15)

In [19]:
data.head() # 58 records, good, but lost plane type, bad
# where's 'extract1' - can we have multiple indexes or have to put that data back in?

split                               Casualties            Circumstances  \
extract                                                                   
Aeroflot Flight 217             Extremely High    Bad Visibility by Day   
Aeroflot Flight 3352            Extremely High  Bad Visibility by Night   
Aeroflot Flight 4227            Extremely High  Bad Visibility by Night   
Aeroflot Flight 7425            Extremely High  Bad Visibility by Night   
Aeroflot/Moldovia (CCCP-65816)  Extremely High  Bad Visibility by Night   

split                              Country Crew        Date Ground  \
extract                                                              
Aeroflot Flight 217                 Russia   10  1972-10-13      0   
Aeroflot Flight 3352                Russia    5  1984-10-11      4   
Aeroflot Flight 4227            Kazakhstan   10  1890-07-08      0   
Aeroflot Flight 7425            Uzbekistan    9  1985-07-10      0   
Aeroflot/Moldovia (CCCP-65816)  Kazakhstan   13  1979-08-11      0   

split                             Latitude               Location   Longitude  \
extract                                                                         
Aeroflot Flight 217              55.755826        Moscow - Russia     37.6173   
Aeroflot Flight 3352            54.9666667          Omsk - Russia  73.3833333   
Aeroflot Flight 4227             43.255058    Almaty - Kazakhstan   76.912628   
Aeroflot Flight 7425            42.1566667  Uchkuduk - Uzbekistan  63.5555556   
Aeroflot/Moldovia (CCCP-65816)     48.8125    Saykyn - Kazakhstan  46.7636111   

split                                    Notes Passengers Phase    Reason  \
extract                                                                     
Aeroflot Flight 217               No survivors        164   APR  Accident   
Aeroflot Flight 3352            Some survivors        169   LDG  Accident   
Aeroflot Flight 4227              No survivors        156   ENR  Accident   
Aeroflot Flight 7425              No survivors        191   ENR  Accident   
Aeroflot/Moldovia (CCCP-65816)    No survivors        165   ENR  Accident   

split                          Total Dead Type  
extract                                         
Aeroflot Flight 217                   174  COM  
Aeroflot Flight 3352                  178  COM  
Aeroflot Flight 4227                  166  COM  
Aeroflot Flight 7425                    0  COM  
Aeroflot/Moldovia (CCCP-65816)        178  COM

In [20]:
df.pivot(['extract','extract1'], 'split', 'split1')

ValueError: Wrong number of items passed 870, placement implies 2

### Problem:

we have two columns we want to pivot on, 'extract' & 'extract1' (AKA flight & plane) but

pivot() can't have multiple indexes...

e.g. df.pivot(index = ['extract','extract1'], 'split', 'split1') # error

pivot_table() can but insists on doing some accounting or aggreagating too, like sum or avg, which we don't need

e.g. pd.pivot_table(df, values='split1', index = ['extract','extract1'], columns='split') # error, no function 

### Solutions? 

use pivot() then force the other column back into the data?? 

or trick pivot_table() into doing some pointless accounting (that adds up to nothing)??

try something else... stack, unstack, group, dplyr?

DIY code??



![](http://www.desktopimages.org/pictures/2014/0212/1/orig_150933.jpg)

In [21]:
# solution: make the function a copy, x = x
# data = pd.pivot_table(df, index=["extract","extract1"], columns = 'split', values = 'split1', aggfunc = lambda x: x)   
# or 
data = pd.pivot_table(df, index=["extract","extract1"], columns = 'split', values = 'split1', aggfunc = 'max') 
# ha, cop that
# http://stackoverflow.com/questions/19279229/pandas-pivot-table-with-non-numeric-values-dataerror-no-numeric-types-to-ag

data # not using df anymore, keep it as backup

split                                                                                Casualties  \
extract                                    extract1                                               
Aeroflot Flight 217                        Ilyushin Il-62                        Extremely High   
Aeroflot Flight 3352                       Tupolev Tu-154                        Extremely High   
Aeroflot Flight 4227                       Tupolev Tu-154B-2                     Extremely High   
Aeroflot Flight 7425                       Tupolev Tu-154B-2                     Extremely High   
Aeroflot/Moldovia (CCCP-65816)             Tupolev Tu-134A(both)                 Extremely High   
African Air (RA-26222)                     Antonov An-32B                        Extremely High   
Air France Flight 447                      Airbus A330-203                       Extremely High   
Air India Flight 182                       Boeing 747-237B                       Extremely High   
Air India Flight 855                       Boeing 747-237B                       Extremely High   
Air New Zealand Flight 901                 McDonnell Douglas DC-10-30            Extremely High   
Alas Nacionales Flight 301                 Boeing 757-225                        Extremely High   
Alia Royal Jordanian (JY-AEE)              Boeing 707-321C                       Extremely High   
All Nippon Airways Flight 58 and           Boeing 727-281 and F-86 Sabre         Extremely High   
American Airlines Flight 11                Boeing 767-223ER                      Extremely High   
American Airlines Flight 191               McDonnell Douglas DC-10-10            Extremely High   
American Airlines Flight 587               Airbus A300B4-605R                    Extremely High   
American Airlines Flight 77                Boeing 757-223                        Extremely High   
American Airlines Flight 965               Boeing Tu-154M                        Extremely High   
Arrow Air Flight 1285                      Douglas DC-8-63CF                     Extremely High   
Avianca Flight 011                         Boeing 747-283B                       Extremely High   
British Airways Flight 476 and             Hawker Siddeley Trident 3B and        Extremely High   
Caspian Airlines Flight 7908               Tupolev Tu-154M                       Extremely High   
China Airlines Flight 140                  Airbus A300B4-622R                    Extremely High   
China Airlines Flight 611                  Boeing 747-209B                       Extremely High   
China Airlines Flight 676                  Airbus A300-622R                      Extremely High   
China Northwest Airlines Flight 2303       Tupolev Tu-154M                       Extremely High   
Cubana de Aviación Flight 9646             Ilyushin II-62M                       Extremely High   
Dana Air Flight 992                        McDonnell Douglas MD-83               Extremely High   
EgyptAir Flight 990                        Boeing 767-366ER                      Extremely High   
Garuda Indonesia Flight 152                Airbus A300B4-220                     Extremely High   
Icelandic Airlines Flight LL 001           Douglas DC-8                          Extremely High   
Inex-Adria Aviopromet Flight 1308          McDonnell Douglas MD-81               Extremely High   
Iran Air Flight 655                        Airbus A300B2-203                     Extremely High   
Iranian Air Force (15-2280)                Ilyushin Il-76MD                      Extremely High   
Japan Airlines Flight 123                  Beoing 747SR-46                       Extremely High   
Kenya Airways Flight 431                   Airbus A310-304                       Extremely High   
Korean Air Flight 801                      Boeing 747-3B5                        Extremely High   
Korean Air Lines Flight 007                Boeing 747-230B                       Extremely High   
LOT Flight 5055                         

### That's most of the wrangling as was done with DataWrangler, there are a few more optional steps:
* want the manufacturer e.g. Beoing?
* remove 'extract' & 'extract1'?
* rename columns ('split') 
* extract year into new column 
* export e.g. df.to_csv(file_name, sep=',')

In [ ]:
#data.reset_index()

# Part 2 

## Wait there's more:

Bad, bad data

This data has been deliberately damaged (sorry)

Some are obvious, some are subtle (some were already there... e.g. look for 'Â')

### See if you can find them

In [ ]:
# start with a summary table
data.describe()

### describe() shows:

* count - we can see that there are 58 records across the board, no surprise (what would it mean if there were non 58s?)
* unique - looks like all the 'Casualties' are identical (unique = 1, i.e. all 'Extremely High'), maybe this column is redundant?
* top - interesting, there were two major disasters in the exact same place? Check dates?
* freq - also interesting, 3 times there were 156 passengers on flights... superstitious? Or is it bad data?

So some clues here, dig deeper:

In [ ]:
#We can also describe individual columns:
data['Crew'].describe()

In [ ]:
# nothing new here, the most common number crew size is 14 (7 times)
# what's the biggest crew?
data['Crew'].max()


Why 9 (or even '9')?

How can this max be less than 14?

Are these even numbers?


In [ ]:
data['Crew'].mean() # expecting ~15, they were all big planes?

This is weird

If they are numbers it should be higher

If they are not numbers, what does mean 'mean'? 

In [ ]:
data[data['Crew'] == 14] # look at all the crew = 14 planes, should be 7

In [ ]:
# there are none... try this:
data[data['Crew'] == '14'] 

In [ ]:
# so there's a clue, 14 vs '14' 
# what are these data types anyway?
data.dtypes

### All data are of type object... 
### Overruled:

In [ ]:
data = data.convert_objects(convert_numeric = True) 
# this is a bit brutal, can you convert when data is loaded?

In [ ]:
data.dtypes

In [ ]:
# so now we have some numbers (int & float)
data.describe()

### Better, more information, describe() now shows:
* count - as above, all 58
* mean - crew ~15, seems OK, but mean lat & long doesn't mean much... or does it?
* std - Standard Deviation
* min - now we see some problems, zero crew? Was this a way to code a hijacking, or is it missing, or should it be 10, 20, 30?
* 25, 50 & 75% are quartiles...
* max - crew 181, no way! 1692 dead, no plane is that big... or could this be Lockerbie, i.e. plane hit town

Let's investigate the crew data:

In [ ]:
data[data['Crew'] > 20] # try also e.g. < 10

So Boeing 747s have large crews...
#### 33 crew is that possible? Or is that two flights?

#### 181 crew? Same value as for 'Total dead', can you derive crew from dead minus passengers?
(or does 'Some survivors' corrupt the maths?)


# Another way to explore... plot that data:


In [ ]:
%matplotlib inline 
# notebook majik to display plots in the notebook

data[['Crew','Passengers']].plot(x= 'Crew',y= 'Passengers',kind= 'scatter');

#### And there's the extreme outlier
#### Are there any others?

In [ ]:
# so how to put a number to an outlier?
data['Crew'].max() # works, now that they are numbers

In [ ]:
data[data['Crew']==data['Crew'].max()]

## Plotting non-numeric data

In [ ]:
data['Phase'].value_counts().plot(kind='bar');
# you can guess some of these codes
# ENR = en route?
# APR = Aproach
# Takeoff, Landing
# ICL?

In [ ]:
# not much point in plotting Casualties... so what the hey
data['Casualties'].value_counts().plot(kind='bar')

# To do: find any other data problems (there are about 10)

<br>

![](http://media1.popsugar-assets.com/files/thumbor/yjoSwHRBZ4MpTO3TN6lvI_gsKMI/fit-in/2048xorig/filters:format_auto-!!-:strip_icc-!!-/2016/03/02/901/n/1922283/01f64bd801c06153_game4/i/When-Everyone-Trying-Talk-You-Youre-Too-Hungry-Care.gif)

#### Post your suspected bad data cases in Moodle discussion forums, how you found it, and suggested fixes
(one or two each, share the load)


In [ ]:
# go crazy

#### Can data be 'typed' as it is read in? 
(yes see 'dtype')
http://pandas.pydata.org/pandas-docs/stable/generated/pandas.read_csv.html
#### What happens if this process encounters bad data?

#### So why is 'Crew' max '9' above?

#### When a plane hits another plane is that one record or two?

#### Can DataWrangler do this sort of wrangling? 

